File: flutter/lib/features/weather_search/data/repositories/weather_repository_impl.dart
import '../../../../core/network/api_client.dart';
import '../../../../core/network/api_configuration.dart';
import '../../domain/entities/weather_exception.dart';
import '../../domain/entities/weather_forecast.dart';
import '../../domain/repositories/weather_repository.dart';
import '../datasources/weather_endpoint.dart';
import '../mappers/weather_forecast_mapper.dart';
import '../models/weather_forecast_response.dart';

/// `WeatherRepository` 的實作：透過注入的 [ApiClient] 取得全部縣市預報、解析，
/// 並將失敗轉為 `WeatherException`。網路細節由 client 負責，本型別只負責組請求、解析與領域對應。
class WeatherRepositoryImpl implements WeatherRepository {
  WeatherRepositoryImpl({required this.client, required this.configuration});

  final ApiClient client;
  final ApiConfiguration configuration;
  static const WeatherForecastMapper _mapper = WeatherForecastMapper();

  @override
  Future<List<WeatherForecast>> fetchAllForecasts() async {
    final dynamic data;
    try {
      // 不帶 locationName → API 回傳全部縣市。
      data = await client.get(
        '${configuration.baseUrl}/${WeatherEndpoint.thirtySixHourForecast}',
        queryParameters: {'Authorization': configuration.authorizationKey},
      );
    } on ApiException catch (e) {
      // 授權失敗（401）單獨處理，方便辨識金鑰問題；其餘傳輸 / 5xx 歸為 requestFailed。
      if (e.type == ApiErrorType.unacceptableStatusCode && e.statusCode == 401) {
        throw const WeatherException(WeatherErrorType.unauthorized);
      }
      throw const WeatherException(WeatherErrorType.requestFailed);
    }

    final WeatherForecastResponse parsed;
    try {
      parsed = WeatherForecastResponse.fromJson(data as Map<String, dynamic>);
    } catch (_) {
      throw const WeatherException(WeatherErrorType.invalidData);
    }

    return parsed.locations.map(_mapper.map).toList();
  }
}

File: ios-native/Packages/Sources/DataModule/WeatherSearch/Repository/WeatherRepositoryImpl.swift
import DomainModule
import Foundation

/// ``WeatherRepositoryProtocol`` 的實作：透過注入的 ``APIClient`` 取得全部縣市預報、解析，
/// 並將失敗轉為 ``WeatherError``。網路細節（送出請求、狀態碼）由 client 負責，
/// 本型別只負責組請求、解碼與領域對應。
public final class WeatherRepositoryImpl: WeatherRepositoryProtocol {
    private let client: APIClient
    private let configuration: APIConfiguration

    /// - Parameters:
    ///   - configuration: base URL 與授權碼（由 App 層注入）。
    ///   - client: 網路實作，預設 ``URLSessionAPIClient``；App / 測試可注入自訂或 mock。
    public init(configuration: APIConfiguration, client: APIClient = URLSessionAPIClient()) {
        self.configuration = configuration
        self.client = client
    }

    public func fetchAllForecasts() async throws -> [WeatherForecastModel] {
        let request = try makeForecastRequest()

        let data: Data
        do {
            data = try await client.data(for: request)
        } catch APIError.unacceptableStatusCode(401) {
            // 授權失敗單獨處理，方便辨識金鑰問題。
            throw WeatherError.unauthorized
        } catch {
            // 其餘傳輸 / 非 2xx（逾時、5xx 等）歸為 requestFailed。
            throw WeatherError.requestFailed
        }

        let decoded: WeatherForecastResponse
        do {
            decoded = try JSONDecoder().decode(WeatherForecastResponse.self, from: data)
        } catch {
            throw WeatherError.invalidData
        }

        return try decoded.records.location.map { try WeatherForecastMapper.map($0) }
    }

    /// 組出「不帶 locationName（取回全部縣市）+ 授權碼」的請求。
    private func makeForecastRequest() throws -> URLRequest {
        let endpointURL = configuration.baseURL.appendingPathComponent(WeatherEndpoint.thirtySixHourForecast)
        guard var components = URLComponents(url: endpointURL, resolvingAgainstBaseURL: false) else {
            throw WeatherError.requestFailed
        }
        components.queryItems = [
            URLQueryItem(name: "Authorization", value: configuration.authorizationKey),
        ]
        guard let url = components.url else {
            throw WeatherError.requestFailed
        }
        return URLRequest(url: url)
    }
}
